### ISSUE TITLE
Dropout(inf) result differs on CPU vs CUDA

### ISSUE THREAD

**Issue by jithunnair-amd** [pytorch/pytorch#59421]
## 🐛 Bug

Applying dropout on a value of Inf results in different outputs on CPU vs CUDA. The CUDA output converts Inf to 0, whereas the CPU output converts Inf to Nan. This is seen on both fp32 and fp16.

## To Reproduce

>>> d = torch.nn.Dropout(p=0.1)
>>> x = torch.full([1000], float("Inf"))   ### CPU FLOAT32 TENSOR WITH EVERY ELEMENT BEING "INF"
>>> torch.isnan(d(x)).any().item()
True
>>> x = torch.full([1000], float("Inf")).cuda()   ### GPU FLOAT32 TENSOR WITH EVERY ELEMENT BEING "INF"
>>> torch.isnan(d(x)).any().item()
False
>>> x = torch.full([1000], float("Inf")).cuda().half()   ### GPU FLOAT16 TENSOR WITH EVERY ELEMENT BEING "INF"
>>> torch.isnan(d(x)).any().item()
False

## Expected behavior

CPU and CUDA results should match

## Environment

Please copy and paste the output from our
Collecting environment information...
PyTorch version: 1.8.1+cu111
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.25

Python version: 3.6 (64-bit runtime)
Python platform: Linux-4.15.0-65-generic-x86_64-with-Ubuntu-18.04-bionic
Is CUDA available: True
CUDA runtime version: 11.1.105
GPU models and configuration:
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB
GPU 4: Tesla V100-SXM2-32GB
GPU 5: Tesla V100-SXM2-32GB
GPU 6: Tesla V100-SXM2-32GB
GPU 7: Tesla V100-SXM2-32GB

Nvidia driver version: 460.73.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] torch==1.8.1+cu111
[pip3] torchaudio==0.8.1
[pip3] torchvision==0.9.1+cu111
[conda] Could not collect

cc @albanD @mruberry @jbschlosser

**Comment by jithunnair-amd** [comment]
cc @jeffdaily @sunway513

**Comment by imaginary-person** [comment]
https://github.com/pytorch/pytorch/blob/c371542efc31b1abfe6f388042aa3ab0cef935f2/aten/src/ATen/native/Dropout.cpp#L65-L67

In the CPU implementation, it looks like `INFINITY` is being multiplied by some 0s in `noise`, and that's resulting in NaNs.

**Comment by jeffdaily** [comment]
It seems the CPU and CUDA implementation both will multiply inf by 0, but the CUDA result is 0 instead of nan.  CUDA code below.

From `fused_dropout_kernel_vec`:
https://github.com/pytorch/pytorch/blob/c769300301e8bc775e54cdd114e2beaa11b1f893/aten/src/ATen/native/cuda/Dropout.cu#L86-L91
https://github.com/pytorch/pytorch/blob/c769300301e8bc775e54cdd114e2beaa11b1f893/aten/src/ATen/native/cuda/Dropout.cu#L102-L107

From `fused_dropout_kernel`:
https://github.com/pytorch/pytorch/blob/c769300301e8bc775e54cdd114e2beaa11b1f893/aten/src/ATen/native/cuda/Dropout.cu#L150-L153
https://github.com/pytorch/pytorch/blob/c769300301e8bc775e54cdd114e2beaa11b1f893/aten/src/ATen/native/cuda/Dropout.cu#L169

**Comment by imaginary-person** [comment]
> It seems the CPU and CUDA implementation both will multiply inf by 0, but the CUDA result is 0 instead of nan. CUDA code below.

Yup, one potential solution might be to write a custom multiply CPU kernel that'd return 0 instead of `Infinity`. It can be called from here - https://github.com/pytorch/pytorch/blob/c371542efc31b1abfe6f388042aa3ab0cef935f2/aten/src/ATen/native/Dropout.cpp#L33 & https://github.com/pytorch/pytorch/blob/c371542efc31b1abfe6f388042aa3ab0cef935f2/aten/src/ATen/native/Dropout.cpp#L39

**Comment by jithunnair-amd** [comment]
IIUC, the CPU operator behaviour is usually what is considered as the reference to match. So maybe the first thing to align on is which behaviour we are going to consider as the "expected" one, and then modify the kernels as needed.

**Comment by imaginary-person** [comment]
> So maybe the first thing to align on is which behaviour we are going to consider as the "expected" one, and then modify the kernels as needed.

Thanks, but the description of `torch.nn.dropout` states that some elements should be zeroed out, so it seems that the CPU implementation should be revised.

**Comment by jeffdaily** [comment]
Should nans also be zeroed?

**Comment by imaginary-person** [comment]
> Should nans also be zeroed?

Hello @albanD, please advise as to what would be the correct thing to do here. Thanks!

**Comment by ngimel** [comment]
I'm getting nan results on master cuda 11 build:
```
In [6]: d = torch.nn.Dropout(p=0.1)

In [7]: x = torch.full([1000], float("Inf")).cuda()

In [8]: torch.isnan(d(x)).any().item()
Out[8]: True

```
but colab indeed gives 0's, that's weird, there have been no substantial changes to dropout since 1.8

**Comment by imaginary-person** [comment]
Colab has PyTorch built with CUDA 10.1

**Comment by albanD** [comment]
> Hello @albanD, please advise as to what would be the correct thing to do here. Thanks!

In general, we allow nans to propagate as much as possible as we don't want to "hide" them without the user knowing.

**Comment by jeffdaily** [comment]
That answers the questions about zeroing nans, thank you @albanD.  But what should the behavior of infs be for dropout?

**Comment by albanD** [comment]
For "inf" I'm not sure tbh.

But I think it is natural to return "nan" here (and it will make the implementation simpler).
Also that would be consistent with the behavior of `tf.keras.layers.Dropout` (that returns nan for both + and - inf) which is always nice.

**Comment by jithunnair-amd** [comment]
@ngimel @albanD Will someone from PyTorch side be looking into this discrepancy wrt the Inf input to torch.nn.Dropout?

**Comment by ngimel** [comment]
Given that cuda 11 builds produce expected `nan` results for `inf` inputs, I don't think we'll be digging deep into cuda 10 failures.

**Comment by jeffdaily** [comment]
Since we're settled on the cuda 10 behavior as erroneous, this issue is a "wontfix".
